docs/adr/2024-08-14_009-refscripts-fee-change.md

```diff
@@ -0,0 +1,78 @@
+
+In order to prevent such an attack a `"minFeeRefScriptCostPerByte"` protocol parameter was introduced in the Conway era. The idea was fairly simple: we would calculate the total size of reference scripts used by a transaction and multiply it by the value specified by this protocol parameter. Result would be added to the transaction fee. However, in order for this approach to be a definite deterrent of such attacks this parameter would have to be set to a fairly high value. Community was not very keen on having this value set to a high enough value, since that could make reference scripts almost as expensive to use as regular scripts, thus significantly reducing their usability. Taking this fact into consideration, a decision was made to set this parameter to a fairly moderate value to at least deter potential attacks like that by at least making it pretty expensive, instead of prohibitively expensive. This would allow us to fine tune the parameter or potentially change the pricing algorithm at a later point after we were already in the Conway era. This reasoning was coming from the fact that at that point we would be able to reveal to a wider audience the true nature of this issue.
+
+Unfortunately things did not go exactly as planned, because on [June 25th 2024 an attack like this actually took place](https://cardanospot.io/news/ddos-attack-on-the-cardano-blockchain-mRIKAzZTNnzq5NGd). This attack forced us to make a quick decision on implementing a definite prevention of such attacks, while trying to not have a drastic impact on the common use case that DApp developers rely on so much.
+
+## Decision
+
+Linear pricing was either too expensive when the multiplier was set too high or was an inadequate deterrent when the multiplier was set too low. Therefore, we needed to implement a pricing mechanism that would be very expensive for usage with large quantities of large plutus scripts, while keeping the pricing reasonably low for the most common use case of a total size of reference scripts of at most 25KiB per transaction. One of the constraints we had to operate under was inability to add any new protocol parameters, since that was a bit too late in the release cycle of the Conway era. In other words we had to hard code some values, which will be turned into proper protocol parameters in the next era.
```

Comment:
> common use case of a total size of reference scripts of at most 25KiB per transaction.

Do you have data that shows the distribution of reference script sizes used in transactions? If so, what does it look like?

Reply:
Some basic stats for all transactions (since their introduction up until slot `132378301`, excluding the txs from the attack) with positive ref scripts size:

 - Number of txs analyzed: 7445529
 - Maximum ref scripts size: 251916B (this [tx](https://cexplorer.io/tx/8086920ca8ec429a57c3aeda411fa44175078843e2ab451cea2b83eb10db153a), seems to be accidental that the txins also contain ref scripts)
 - Mean ref scripts size: 5362B
 - A few percentiles:
    - Median: 2561B
    - 90%: 14174B
    - 95%: 16480B
    - 99.99%: 23022B

<details>
<summary>
Largest 100 txs by ref scripts size
</summary>

```
84771377	8086920ca8ec429a57c3aeda411fa44175078843e2ab451cea2b83eb10db153a	251916
80232496	2e70b7a39059b0b79afe33f5443466dfa8da71e332487e15aa11abfd8c397ca3	94992
121910776	b6c88753597872cf7ad5e1d2ff2ca02a6a31b3cc63980829e01f40e1abeb047a	88096
81116763	0df03c6b649754ff963d4ef942e0abe1c7b8e5d391800123d9d657b01d88bc33	83972
81117712	bed62b4d2d72f43f582b0c8ab8e8202eeefa170bfba58086be0e15ea92310abe	83972
94513305	b51313dd65c8c0c1bc132a0b85fc3b067b7bf2fee82a16f324dd58f30a99ac1c	83972
81115343	8312984035c6b46ffba36185d3d2f5e999e6541a72f63cb5a8c4a12f7721ed0f	83969
94511437	b40db2e28ca084b8f99f9a90f8c528e46b9db954dc231bbe31d06f84e37fe65f	76660
80231345	0bc7051186ec125296b08416cfc0829db7c8b007985e52438c99056ab131e9d6	74937
94509293	20ca05dd6a16ff7f8ac3d5d3b7a4bbbbad9c2bcf10b5742abf86729544d18bdc	72481
72399352	42c879180340f05562b0f252c19cb627a9c27969dfefb0279906222d7bc509b1	57877
117026041	c818c635a773615a68455dd173051675a1d9b0f64cce7a88e73231c3df966085	51141
103469938	4e9e9f309006433412f8d82bd51327104e93c24112c57f02194d30c60df5e7f6	51124
102282497	e2a6c1e731bc4dc573459ff077b62723748dd12b6ff7389819179928893a5377	51112
122777351	ae4219238743d0b201355cf6ba2b62b444ad324953f7a3e4d7442414c9f220f2	42550
122779350	e6034686c50d7764b9a182db86ef0346a1e8876547ff7864bf1f45ad2818c6a0	42550
124671238	0cad5c7e2fb9d5c79cec056e622432a86ca4cf6a597c8e28684c6e030d3221dd	42550
124723789	fda9567078bb382af852b62950112657a5f806317e2950d31b01a9677c1e2ffa	42550
124724761	8d827e84a2c844f57279f57878a006078c796bff3a38e3c6a3ac63c7f7ad736a	42550
124725993	dfc78409270277e30ad21f8e36c28228dca214b8fa21ec716723248db8c821cd	42550
124726049	4766436944245a71cdd426778a6985a3592cbfd4412607ab10691f7b8f9fcf11	42550
124760039	ca48da1b93d74241909f3f4caa0c652bbb929ac55676d8a18a1c2739daecda64	42550
124891334	e4144d2dcbf0ee67756d96f9d45ca2099d94865c5a54a3215c508e9aa4628b0d	42550
124892861	ef368e6e9bf877bcbbf66ab1028fce48b318eae68e5f0ff10b50d3bc8e0a4821	42550
124898708	3240941ef43202db17754a7c1f14138833e6125a97cc2c2188a52f64f9689e31	42550
124898760	6678434363643dd999e1206625f2fdffc2a67353c18289a2a6f7581488b9e1bf	42550
124899976	98f6acb88bfbd0b8cfe9a79f0f21f1c9a76552526234edbb0bd56c0faa3acaf2	42550
124914540	fa0395e03752570e756a073231b975475349a562325c9082dba309cfbb9992e8	42550
124920268	2e139a877b5a5c2223e3fb726325edf81e6dcb3c72ac7e8e2764b67eed0165fa	42550
124926610	dfa67eac7ab2c377d90d7a43b615cbf13fa799dca36f588e25f900d1f28a8849	42550
124932179	86fbd6339e97fc99259576fd67ee19fadc41857f75006952e995ceec1d26ff75	42550
125001151	4ed689eefba747d1083361c83ed662a2e853acdd6d62c8f5c13aca1c970e5dff	42550
125223913	2d7e51bfdb7d19f1594a71cebb12f62c1f978ccd4183a93c6b91c5db38a22b2c	42550
125226657	ad13287856d76b20928115aeb104a92fe0902284fc430a76e71ca942aa4919fe	42550
125484089	f16f6e9c5fabc18c743492d6c0116fe7778a649a78bd31b6eed0a99549d76187	42550
125484743	2b134bda02f1bcd47755d1cad933d881a46709d0e2aeee6f854811eb19d35efb	42550
125513806	aa829e6fbe1aa6f0f3466b43fb87acbcaeef438812a9603290f708c2ece80265	42550
126868405	05c3c44961b69947f77b7c5b087b872b596803b5accff67eceed57e27b4cf8bd	42550
126886324	70a3b051ec8cd331a0b833bf4e4745c9da731846ffb5a499a68bd2d90da658aa	42550
126893523	94adf69c34be9eb0f3fe16cd46c0f8adfe7300b3ad99aec499c5987d486c9842	42550
127075042	f285dc671eee6554aeecd0af4af38636a4ad2b62916d0ececae16eaeab4d7c40	42550
127109964	6fd72f821d79e9a804cad0f060adbd9defe5bb61b49233138e4aa83c3a50c56d	42550
127444362	bc038c25b48ba693733c55408348261c15784594534a07a625b76fe0afd0b1ff	42550
127449835	b12dc90b9d31a24e8c8ba016d8433d9a1b84baf983a92815094c694f22b62b09	42550
128832929	e4a5fb33684d10a4c7ee488d09cae8f33b0dbde02aeab0f2a256fc578ba111ca	42550
129750149	5611349b17dc34b1ae8bfc021790f0aeed1eb59c9b0ab3c48333fb7e242a4d0f	42550
130174326	61d3bddaf1885aece395b49b5bf25cf93d8a463b2358a307a98b87c618eb0e1f	42550
130268521	58fbdbfa248d8de6624ae3621171408732ba699b0aef905baec1ff2662131f88	42550
103229302	7809893bb6f90e872b0ec6ca20b7246d815208bf7a58bacd9545240a9e710734	40139
116985583	db4295ddfd0936c92170278103699c3def79d67bc5e37e972223be8fce6467a5	40133
117643134	1b4b2b721efeb9acb9e46806b98573ad97dc124a8fa4c5053e8658a224af4911	40125
102334543	3b82950f6c064f35e53850a756a434ea69aa77c7b4e13dac1e182cb6f41a8555	40123
126547979	2be823d13455ae80d38c7d12203e696d736609637d4af6fa10a8aaea2b3716dc	38677
128582066	5e3cfd287e0cd5531d563633e83eef442e5bab837a3e7dbcbfd6f414ba02688e	38677
101831004	5eed82116887d5fab4e2dd42b954de71d0af473c613aef2f0be710b65a829c97	37805
120614349	5067bc4f654db9227fed4494d983a702336be7acfc94a399e4c799332bba0081	37532
103227952	2c5cbd515fef69b2a1c9dd98e84b0d628fb4685ca3c449d2b5c192f34b99b22a	37449
103227995	8dae15e066cb6c62e0d485ed083f91f3ac6d110e169d32de5caff7885901897f	37449
103228026	d50cf6407e28dd983c0ff7d4265dd5bbcc75f9970522ffdf96a04dab926e9f49	37449
103228056	864c6d41f3240e3c1bf63897f552244790327aeb4f0f401d23a9718fcafde533	37449
103228097	34087726f4a7c877e6525b76a89dfb0e6136c8afb52a5b34c2b0338cd67d97af	37449
103228114	8ebae198c6fe6415415d37b00ff7ab127929bc6c435a56901833bc9b7fb4d765	37449
103228284	6230ea9d9f15543f104bd81c8c5360e76b8f8c3616dd913a59bac3c6f9ad9186	37449
103228321	5bee6b9ea96ecdd72962be3af02b2a0a4d25acbbc92f3f756cc88750e25338f2	37449
103228387	53c6314ccc8eca0ba013412113debc64f9c01e62b0e7c8106a9700381bdf0ecd	37449
103228477	9b223aa2f63b839ebf4f5524b38565c1a4748120a8ae9a4f3e96da0538b33d41	37449
103228641	c2226d5d7b81ca1666d9ab967fc97eac7ce5e269e5d4bc439d13598b541eb1d4	37449
103228718	b8d58dcd3f73cb5da38956fa5f6ec2833a0dd9e4b0fed7405170f9a0bfefce1d	37449
103228828	c950a4dc1502af5100db29f6f9b7c3b814ed082e461f73d368e6defd3b9b6472	37449
103228896	141f377cc002725ecc06497d4783fb633be00df1aa15dd3049fc0b55fb1bbf9b	37449
103228992	c30b3bfb41352257f68fd1b2dc9fa4013957abf20b7536a3851f3d2630b8802c	37449
103229015	6bf7eaa885310cdcd10d1e141771f3d53d8c2841e31e337d304b079a7aaaf74d	37449
103229057	598544d5a0c92f5aab04d11fc8e1755e812ead334a74e898af1280c391be77ce	37449
103229083	a208be2887cd25a52f61bb2580d36f9cab4ad493803bdb640c5794d33c9d7859	37449
103229181	cdde9c8fc14fb9e28a85ad020feb65f606a39cf3a728117b0c4cebc720531566	37449
103229256	af5730d4c09f7885fa4dc970ebdfadb5c6f1447acdd7f8b6f5e54528f0920e9b	37449
116985481	42a20b4f2190e6d21893c6d8c92d67f01f6dca5f1bb172dd80d39d9699d00c75	37443
116985513	82516b99e7886537ee816c36199f5049a36debbf04ca7fba5d67d0429e8538a0	37443
116985536	451d5d8b4000b1ceb98265849ac8f5cefb7a0d76edf4a9736dcd568d85ce60f8	37443
117642562	e1bb46e3867107b4d9afeab766930bbddf5e8699f989ac260b0876b1fe6703d0	37435
117642606	0945efad80f49767240a0683c932a9b94c448418e2316d193545eeac84a10462	37435
117642634	58073d78f94c5c1f0c722800858d3aa56921dcba4e94f02a034ae167679c80ea	37435
117642745	60fc594b8789d6f79cc0e2bcc2ad7683514c25274b837f0365136c3a623e2bac	37435
117642813	ee4b5456933fb42b6ca78e3d4b69542731f4c7f4d39c9621907a59808ffd773c	37435
117642860	1f50358d58ce5bc4a60d31795b7a6d88311d1c753b7ad2e5e2a6bb8a356e7816	37435
117642916	97ae58584372c2f7e664b49aca9adddaf6271ee32eea81c74189e26182770d80	37435
117642961	18bb89d5f7f37fb8e59c464e51880122165cb7131d99e04ae13b88b60374807f	37435
117643083	ca39f0cdc4d3a2d58f39317357b26c69273efecc100fbd45a77e5d856a8a2e0b	37435
102333862	5072333975bbe34164cff36208c59661d844785897ca5b33519d824331b7658f	37433
102333959	58d780cb6ef4b3cd336b9f08ef6e1237ea28431d6600db2643db796b795f54da	37433
102334064	f34870967511760aa83e686ee3f9853c11cdc3c230791e63c6be1863afaaa7eb	37433
102334209	55ecb5dcaa28b7ad747c4967b607f7b2ebd7abb94f0f3aa876f2eb7fafe16a01	37433
102334270	0b8be0c5046bcf3daaf796cece4df94f5c4e931b11eb0c36412c19b263092b8d	37433
102334379	9cae3c43af0b36f798767c35dbba71e294d99337392374b2799f0746099825a1	37433
122976693	cf4ca98c804e770729847d2a05052ce1f244cf504f0c26537baa193b87a96ed9	36541
116194749	b51f91cab20fefae6a5a62aac70aae8f51fc9433d50c0ece2c6964d9464bd624	36280
116218672	bbf6cbd58ea7ac4d4381dca22c605e8da31999269641db1af9733dd9f2343c1f	36280
116218756	84b832d85228f21d443617f90f87793e185221505e916cca83b08f0d38f72218	36280
116222386	fd14bf6d74a46fec8f5d2b1edff559ffbd936c8153e299d017e21a7356fac4c4	36280
116230860	fb7e40cc20f52bcd64db62a47470597587cf037bf54c020683473fa7b5c2cfc5	36280
```

</details>